core/payment_channel_manager/impl/payment_channel_manager_impl.cpp

```diff
@@ -0,0 +1,310 @@
+    OUTCOME_TRY(message_state, message_wait.waitSync());
+    if (message_state.receipt.exit_code != VMExitCode::Ok) {
+      return PaymentChannelManagerError::kSendFundsErrored;
+    }
+    return std::move(message_cid);
+  }
+
+  outcome::result<AddChannelInfo>
+  PaymentChannelManagerImpl::createPaymentChannelActor(
+      const Address &client, const Address &miner, const TokenAmount &amount) {
+    // payment channel constructor params
+    PaymentChannelConstruct::Params construct_params{.from = client,
+                                                     .to = miner};
+    OUTCOME_TRY(encoded_construct_params,
+                codec::cbor::encode(construct_params));
+
+    // init actor exec call params
+    InitActorExec::Params init_params{
+        .code = kPaymentChannelCodeCid,
+        .params = MethodParams{encoded_construct_params}};
+    OUTCOME_TRY(encoded_init_params, codec::cbor::encode(init_params));
+
+    UnsignedMessage unsigned_message{kMessageVersion,
+                                     kInitAddress,
+                                     client,
+                                     {},
+                                     amount,
+                                     kDefaultGasPrice,
+                                     kDefaultGasLimit,
+                                     InitActorExec::Number,
+                                     MethodParams{encoded_init_params}};
+    OUTCOME_TRY(signed_message, api_->MpoolPushMessage(unsigned_message));
+    OUTCOME_TRY(message_cid, vm::message::cid(signed_message));
+    OUTCOME_TRY(message_wait, api_->StateWaitMsg(message_cid));
+    OUTCOME_TRY(message_state, message_wait.waitSync());
+    if (message_state.receipt.exit_code != VMExitCode::Ok) {
+      return PaymentChannelManagerError::kCreateChannelActorErrored;
+    }
+    OUTCOME_TRY(ret,
+                codec::cbor::decode<InitActorExec::Result>(
+                    message_state.receipt.return_value));
+    Address channel_actor_address = ret.robust_address;
+    std::unique_lock lock(channels_mutex_);
+    saveChannel(channel_actor_address, client, miner);
+
+    return AddChannelInfo{channel_actor_address, message_cid};
+  }
+
+  outcome::result<PaymentChannelState>
+  PaymentChannelManagerImpl::loadPaymentChannelActorState(
+      const Address &channel_address) const {
+    OUTCOME_TRY(chain_head, api_->ChainHead());
+    OUTCOME_TRY(tipset_key, chain_head.makeKey());
+    OUTCOME_TRY(tipset, api_->ChainGetTipSet(tipset_key));
+    auto state_tree =
+        std::make_shared<StateTreeImpl>(ipld_, tipset.getParentStateRoot());
+    return state_tree->state<PaymentChannelState>(channel_address);
+  }
+
+  outcome::result<uint64_t> PaymentChannelManagerImpl::getNextNonce(
+      const ChannelInfo &channel, const LaneId &lane) const {
+    auto lookup_lane = channel.vouchers.find(lane);
+    if (lookup_lane == channel.vouchers.end()) {
+      return PaymentChannelManagerError::kChannelNotFound;
```

Comment:
No vouchers should return nonce=0

Reply:
I suppose you misunderstood, `vouchers` is a lane here. Renamed `vouchers` to `lanes`. Getting nonce of non allocated lane for adding voucher for the lane should be treated as an error imo. Otherwise method `AllocateLane` has no sense. Moreover, from lotus it should be 1:
`
var maxnonce uint64
// iterate for lanes
return maxnonce + 1, nil
`

Reply:
There is an allocation of lane on demand in lotus. Changed behaviour to lotus. Seems that `allocateLane()` method is redundant.